docs/proposals/20191017-kubeadm-based-control-plane.md

```diff
@@ -357,23 +367,107 @@ spec:
+to 3 because 3 votes are required for a majority of 4 members. Since the quorum increased, this extra member buys nothing
+in terms of fault tolerance; the cluster is still one node failure away from being unrecoverable._
+
+_Additionally, that new member is risky because it may turn out to be misconfigured or incapable of joining the cluster.
+In that case, there's no way to recover quorum because the cluster has two members down and two members up, but needs
+three votes to change membership to undo the botched membership addition. etcd will by default reject member
+add attempts that could take down the cluster in this manner._
+
+_On the other hand, if the downed member is removed from cluster membership first, the number of members becomes 2 and
+the quorum remains at 2. Following that removal by adding a new member will also keep the quorum steady at 2. So, even
+if the new node can't be brought up, it's still possible to remove the new member through quorum on the remaining live members._
+
+As a consequence KCP remediation should remove unhealthy machine first and then add its replacement.
+
+Additionally, in order to make this approach more robust, KCP will test each etcd member for responsiveness before
+using the Status endpoint and determine as best as possible that we would not exceed failure tolerance by removing a machine.
+This should ensure we are taking actions in case there are other etcd members not properly working on machines
+not (yet) marked for remediation by the MachineHealthCheck.
+
+###### Scenario 1: Three replicas, one machine marked for remediation
+If an MHC will mark one machine for remediation in a control-plane with three replicas, we will look at the etcd
+status of each machine to determine if we have at most one failed member. Assuming the etcd cluster is still all healthy, 
+or the only unresponsive member is the one to be remediated, we will scale down the machine that failed the MHC and
+then scale up a new machine to replace it.
+
+###### Scenario 2: Three replicas, two machines marked for remediation
+If an MHC will mark two machines for remediation in a control-plane with three replicas, remediation might happen
+depending on the status of the etcd members on the three replicas.
+
+As long as we continue to only have at most one unhealthy etcd member, we will scale down an unhealthy machine, 
+wait for it to provision and join the cluster, and then scale down the other machine.
+
+However, if more than one etcd member is unhealthy, remediation would not happen and manual intervention would be required
+to fix the unhealthy machine.
+
+###### Scenario 3: Three replicas, one unresponsive etcd member, one (different) unhealthy machine
+It is possible to have a scenario where a different machine than the one that failed the MHC has an unresponsive etcd. 
+In this scenario, remediation would not happen and manual intervention would be required to fix the unhealthy machine.
```

Comment:
This makes sense, although we might want to revisit it in the future to allow to get to scenario 2

Reply:
agreed, see https://github.com/kubernetes-sigs/cluster-api/pull/3676#discussion_r492975642